src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeReader.java

```diff
@@ -0,0 +1,455 @@
+            this.postingsSummaryInput = postingsSummaryInput;
+            this.index = index;
+            this.listener = listener;
+            this.context = context;
+        }
+
+        public PostingList execute()
+        {
+            try
+            {
+                PriorityQueue<PeekablePostingList> postingLists = new PriorityQueue<>(100, COMPARATOR);
+
+                executeInternal(postingLists);
+
+                FileUtils.closeQuietly(treeInput);
+
+                return mergePostings(postingLists);
+            }
+            catch (Throwable t)
+            {
+                if (!(t instanceof QueryCancelledException))
+                    logger.error(indexContext.logMessage("Balanced tree intersection failed on {}"), treeIndexFile.path(), t);
+
+                closeOnException();
+                throw Throwables.cleaned(t);
+            }
+        }
+
+        protected void executeInternal(final PriorityQueue<PeekablePostingList> postingLists) throws IOException
+        {
+            collectPostingLists(postingLists);
+        }
+
+        protected void closeOnException()
+        {
+            FileUtils.closeQuietly(treeInput);
+            FileUtils.closeQuietly(postingsInput);
+            FileUtils.closeQuietly(postingsSummaryInput);
+        }
+
+        protected PostingList mergePostings(PriorityQueue<PeekablePostingList> postingLists)
+        {
+            final long elapsedMicros = queryExecutionTimer.stop().elapsed(TimeUnit.MICROSECONDS);
+
+            listener.onIntersectionComplete(elapsedMicros, TimeUnit.MICROSECONDS);
+            listener.postingListsHit(postingLists.size());
+
+            if (postingLists.isEmpty())
+            {
+                FileUtils.closeQuietly(postingsInput);
+                FileUtils.closeQuietly(postingsSummaryInput);
+                return null;
+            }
+            else
+            {
+                if (logger.isTraceEnabled())
+                    logger.trace(indexContext.logMessage("[{}] Intersection completed in {} microseconds. {} leaf and internal posting lists hit."),
+                                 treeIndexFile.path(), elapsedMicros, postingLists.size());
+                return MergePostingList.merge(postingLists, () -> FileUtils.close(postingsInput, postingsSummaryInput));
+            }
+        }
+
+        private void collectPostingLists(PriorityQueue<PeekablePostingList> postingLists) throws IOException
+        {
+            context.checkpoint();
+
+            final int nodeID = index.getNodeID();
+
+            // if there is pre-built posting for entire subtree
+            if (postingsIndex.exists(nodeID))
+            {
+                postingLists.add(initPostingReader(postingsIndex.getPostingsFilePointer(nodeID)));
+                return;
+            }
+
+            Preconditions.checkState(!index.isLeafNode(), "Leaf node %s does not have balanced tree postings.", index.getNodeID());
```

Comment:
nit: Not sure if `CorruptIndexException` or `ISE` makes more sense here.

Reply:
I think these should be `CorruptIndexException`. It is the only reason we are going to get this.